esrally/telemetry.py

```diff
@@ -1926,3 +1927,104 @@ def detach_from_node(self, node, running):
+            May optionally specify:
+            ``master-node-stats-sample-interval``: An integer controlling the interval, in seconds,
+            between collecting samples. Default: 30s.
+        :param clients: A dict of clients to all clusters.
+        :param metrics_store: The configured metrics store we write to.
+        """
+        super().__init__()
+
+        self.telemetry_params = telemetry_params
+        self.clients = clients
+        self.specified_cluster_names = self.clients.keys()
+        self.sample_interval = telemetry_params.get("master-node-stats-sample-interval", 30)
+        if self.sample_interval <= 0:
+            raise exceptions.SystemSetupError(
+                f"The telemetry parameter 'master-node-stats-sample-interval' must be greater than zero " f"but was {self.sample_interval}."
+            )
+        self.metrics_store = metrics_store
+        self.samplers = []
+
+    def on_benchmark_start(self):
+        for cluster_name in self.specified_cluster_names:
+            recorder = MasterNodeStatsRecorder(
+                cluster_name,
+                self.clients[cluster_name],
+                self.metrics_store,
+                self.sample_interval,
+            )
+            recorder.record()
+            sampler = SamplerThread(recorder)
+            self.samplers.append(sampler)
+            sampler.daemon = True
+            # we don't require starting recorders precisely at the same time
+            sampler.start()
+
+    def on_benchmark_stop(self):
+        if self.samplers:
+            for sampler in self.samplers:
+                sampler.recorder.record()
+                sampler.finish()
+
+
+class MasterNodeStatsRecorder:
+    """
+    Collects and pushes the current master node name for the specified cluster to the metric store.
+    """
+
+    def __init__(self, cluster_name, client, metrics_store, sample_interval):
+        """
+        :param cluster_name: The cluster_name that the client connects to, as specified in target.hosts.
+        :param client: The Elasticsearch client for this cluster.
+        :param metrics_store: The configured metrics store we write to.
+        :param sample_interval: An integer controlling the interval, in seconds, between collecting samples.
+        """
+
+        self.cluster_name = cluster_name
+        self.client = client
+        self.metrics_store = metrics_store
+        self.sample_interval = sample_interval
+        self.logger = logging.getLogger(__name__)
+
+    def __str__(self):
+        return "master node stats"
+
+    def record(self):
+        """
+        Collect master node name and push to metrics store.
+        """
+        # pylint: disable=import-outside-toplevel
+        import elasticsearch
+
+        try:
+            state = self.client.cluster.state(metric="master_node")
+            info = self.client.nodes.info(node_id=state["master_node"], metric="os")
```

Comment:
I don't care about the `os` metrics here, in fact the name is included in all cases, but there's no documented way to ask for "no metrics", and `os` seemed cheap to retrieve.

Reply:
I think the `cat` API is the way to go here IMO. But in any case if we decide not to use the `cat` API I wonder if it's possible to retrieve the node IDs from existing metrics store metadata rather than making regular API calls? 

e.g. we collect some of this upfront here https://github.com/elastic/rally/blob/4355771a7d9fa027b83331afc2d5e669942b9a0e/esrally/telemetry.py#L1561-L1571)

Reply:
Resolving as I switched to the cat API. (For what it's worth, `add_metadata_for_node` seems to only be called at the end for the moment, but I haven't explored further.)